### ISSUE TITLE
Deadlock on @synchronized in -[UAChannelRegistrar registerWithChannelID:channelLocation:withPayload:forcefully:]

### ISSUE THREAD

**Issue by brblakley** [urbanairship/ios-library#137]
Every once in a while when I build and run my iOS app that is using UA’s iOS SDK 8.4.0 from Xcode, my app’s UI will display and then a moment later the entire app’s UI will become unresponsive (i.e. tapping buttons, etc. doesn’t do anything). If I pause execution, I can see that it’s locked on the `@synchronized` line in `-[UAChannelRegistrar registerWithChannelID:channelLocation:withPayload:forcefully:]`. See below for the stack trace.

I’ve also seen this happen in TestFlight builds (though there I can’t inspect what the cause is but the user-facing experience is exactly the same so I suspect it’s the same issue).

```
#0	0x000000018c090e64 in __psynch_mutexwait ()
#1	0x000000018c15cb8c in _pthread_mutex_lock_wait ()
#2	0x000000018c15cd94 in _pthread_mutex_lock_slow ()
#3	0x000000018bb046f8 in recursive_mutex_tt<false>::lock() ()
#4	0x000000018bb07f28 in objc_sync_enter ()
#5	0x0000000101750670 in -[UAChannelRegistrar registerWithChannelID:channelLocation:withPayload:forcefully:] at /path/to/my/app/Pods/UrbanAirship-iOS-SDK/AirshipKit/AirshipKit/common/UAChannelRegistrar.m:32
#6	0x000000010179b994 in -[UAPush updateChannelRegistrationForcefully:] at /path/to/my/app/Pods/UrbanAirship-iOS-SDK/AirshipKit/AirshipKit/common/UAPush.m:796
#7	0x000000010179ac10 in -[UAPush application:didRegisterForRemoteNotificationsWithDeviceToken:] at /path/to/my/app/Pods/UrbanAirship-iOS-SDK/AirshipKit/AirshipKit/common/UAPush.m:664
#8	0x000000010173d980 in +[UAAppIntegration application:didRegisterForRemoteNotificationsWithDeviceToken:] at /path/to/my/app/Pods/UrbanAirship-iOS-SDK/AirshipKit/AirshipKit/common/UAAppIntegration.m:47
#9	0x0000000101741820 in ApplicationDidRegisterForRemoteNotificationsWithDeviceToken at /path/to/my/app/Pods/UrbanAirship-iOS-SDK/AirshipKit/AirshipKit/common/UAAutoIntegration.m:358
#10	0x0000000103f7da50 in _dispatch_call_block_and_release ()
#11	0x0000000103f7da10 in _dispatch_client_callout ()
#12	0x0000000103f82b78 in _dispatch_main_queue_callback_4CF ()
#13	0x000000018d0450c8 in __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ ()
#14	0x000000018d042ce4 in __CFRunLoopRun ()
#15	0x000000018cf72da4 in CFRunLoopRunSpecific ()
#16	0x000000018e9dc074 in GSEventRunModal ()
#17	0x000000019322d058 in UIApplicationMain ()
#18	0x00000001001254c8 in main at /path/to/my/app/MyApp/Code/ApplicationDelegate.swift:15
#19	0x000000018bf8159c in start ()
Enqueued from com.apple.usernotifications.UNUserNotificationServiceConnection (Thread 0)Queue : com.apple.usernotifications.UNUserNotificationServiceConnection (serial)
#0	0x0000000103f89304 in _dispatch_queue_push ()
#1	0x0000000193463930 in -[UIApplication remoteNotificationRegistrar:didReceiveDeviceToken:] ()
#2	0x00000001a75fb644 in -[UNRemoteNotificationRegistrar didReceiveDeviceToken:forBundleIdentifier:] ()
#3	0x00000001a760fd64 in -[UNUserNotificationServiceConnection _queue_didReceiveDeviceToken:forBundleIdentifier:] ()
#4	0x0000000103f7da50 in _dispatch_call_block_and_release ()
#5	0x0000000103f7da10 in _dispatch_client_callout ()
#6	0x0000000103f8b2e8 in _dispatch_queue_serial_drain ()
#7	0x0000000103f81634 in _dispatch_queue_invoke ()
#8	0x0000000103f8b810 in _dispatch_queue_override_invoke ()
#9	0x0000000103f8d630 in _dispatch_root_queue_drain ()
#10	0x0000000103f8d39c in _dispatch_worker_thread3 ()
#11	0x000000018c157100 in _pthread_wqthread ()
#12	0x000000018c156cac in start_wqthread ()
Enqueued from com.apple.NSXPCConnection.m-user.com.apple.usernotifications.usernotificationservice (Thread 3)Queue : com.apple.NSXPCConnection.m-user.com.apple.usernotifications.usernotificationservice (serial)
#0	0x0000000103f89304 in _dispatch_queue_push ()
#1	0x00000001a760f364 in -[UNUserNotificationServiceConnection didReceiveDeviceToken:forBundleIdentifier:] ()
#2	0x000000018d09ce80 in __invoking___ ()
#3	0x000000018cf922c4 in -[NSInvocation invoke] ()
#4	0x000000018dc95d18 in __NSXPCCONNECTION_IS_CALLING_OUT_TO_EXPORTED_OBJECT__ ()
#5	0x000000018dafdd54 in -[NSXPCConnection _decodeAndInvokeMessageWithData:] ()
#6	0x000000018da87e00 in message_handler ()
#7	0x000000018c18b854 in _xpc_connection_call_event_handler ()
#8	0x000000018c189360 in _xpc_connection_mach_event ()
#9	0x0000000103f7dae0 in _dispatch_client_callout4 ()
#10	0x0000000103f819d8 in _dispatch_mach_msg_invoke ()
#11	0x0000000103f8b128 in _dispatch_queue_serial_drain ()
#12	0x0000000103f80da0 in _dispatch_mach_invoke ()
#13	0x0000000103f8b128 in _dispatch_queue_serial_drain ()
#14	0x0000000103f81634 in _dispatch_queue_invoke ()
#15	0x0000000103f8c358 in _dispatch_root_queue_drain_deferred_item ()
#16	0x0000000103f9457c in _dispatch_kevent_worker_thread ()
#17	0x000000018c156fbc in _pthread_wqthread ()
#18	0x000000018c156cac in start_wqthread ()
```

**Comment by robinkunde** [comment]
I've come across the same issue, and came here to file it. I can provide some context on what is causing it.

UAChannelRegistrar uses @synchronized to protect itself while its mutating its own state. Under certain conditions, it can cause a synchronous call into the main thread while holding the lock, and the main thread might try to acquire the lock itself, causing this deadlock.

Example:
1. [UAChannelRegistrar succeededWithPayload:] is called on a background thread and acquires the lock
2. From here, it calls [UAPush registrationSucceededWithPayload:] via delegate
3. This in turn calls [UAPush createChannelPayload]
4. This in turn calls [UAPush userPushNotificationsAllowed]
5. Since we're on a background thread, this method makes a synchronous call to the main thread via dispatch_sync(dispatch_get_main_queue(), ^{})
6. This enqueues a block on the main thread and waits until that block is run.
7. If any waiting main thread blocks try to acquire the UAChannelRegistrar lock, a deadlock occurs. In my case, it happens by calling [UAPush updateRegistration].

I've thought a bit about how to resolve this issue without having to refactor UAPush:
1. Just read [UIApplication sharedApplication].isRegisteredForRemoteNotifications without calling into the main thread. It's a property and *should* be fine. (Not ideal)
2. Make everything [UAPush registrationSucceededWithPayload:] does asynchronous. (Not sure what the impact of this would be yet)
3. Move the delegate call in [UAChannelRegistrar succeededWithPayload:] out of the lock. I think this is the best option, and something similar should be done in [UAChannelRegistrar failedWithPayload:].

**Comment by robinkunde** [comment]
Looks like there's already a PR that will address this in a different way: https://github.com/urbanairship/ios-library/pull/136

**Comment by rlepinski** [comment]
@robinkunde Thanks for the detailed explanation. We will try to get a fix out today or tomorrow.

**Comment by emablekos** [comment]
Our QA group started experiencing this in our production app after updating to UA 8.4.0. The app would hang on launch.

It was not reproducible on all devices.

We rolled back UA to 8.3.3 which our previous build was on, and QA is no longer reporting the issue.

**Comment by rlepinski** [comment]
Fixed in 8.4.2.

Thanks @robinkunde and @brblakley for the detailed report
